File: LinkedList/2. Add two Numbers/add-two-numbers.js
/**
 * #2
 * Medium
 * 
 * You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.

#1
  1)convert linkted list into number:
  push node value into an array  2 4  3  => 2 4 3 
  revert the array 3 4 2 

  O(n) => iterate over list, for each node add it before the num string -> "342"  
  convert it to number -> 342

   5 6 4 => 465 -> int
  2) add two numbers
   342 + 465 = 807 
  3) convert final number into linked list
O(n) => convert number into string
iterate over the string from the end 
add each value to the linked list and convert to number

  convert number into array => 7 0 8
  convert array into linked list =>
   807 => 8 0 7 => 7 -> 0 -> 8
#2

iterate over both lists
add node values together
if sum is greater than 10 
save the remainder of sum / 10 and add it to next node
  if next node is null, add it to 0 and add to the list


 * 
 * 
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function ListNode(val) {
  this.val = val;
  this.next = null;
}

var addTwoNumbers = function(l1, l2) {
  //output: new list
  //edge case: different length between l1 and l2, 
  //           one is empty list(X)
  let newList = new ListNode(0); //return newList.next
  let curNewLst = newList;  //use it as current pointer
  let curL1 = l1;  //l1 pointer
  let curL2 = l2;  //l2 pointer
  let carry = 0;
// iterate over both lists
  while(curL1 !== null || curL2 !== null) {
    // add node values together
    let sum = (( curL1 !== null) ? curL1.val : 0) + ( (curL2 !== null) ? curL2.val : 0) + carry;
    carry = Math.floor(sum / 10);
    curNewLst.next = new ListNode( Math.floor(sum % 10));
    curNewLst = curNewLst.next;
    if (curL1 !== null) curL1 = curL1.next;
    if (curL2 !== null) curL2 = curL2.next;
}
// if sum is greater than 10 => if carry > 0
// save the remainder of sum / 10 and add it to next node
//   if next node is null, add it to 0 and add to the list
  if (carry > 0) {
    curNewLst.next = new ListNode(carry);
  }
  return newList.next;
};

File: leetcodes/2. Add two Numbers/add-two-numbers.java

// public class ListNode {
//   int val;
//   ListNode next;
//   ListNode() {}
//   ListNode(int val) { this.val = val; }
//   ListNode(int val, ListNode next) { this.val = val; this.next = next; 
//   }
public class Solution {

  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    // input : ListNode l1, ListNode L2
    // output ListNode result
    // constrain: no leading zero,
    // edge case: the length of list in [1, 100], value in [0,9]
    /***
     * test cases: l1 = [0,1] l2 = [0,1,2] one list is longer l1 = [1] , l2 = [1] no
     * empty list, same length of both lists l1 = [9, 9] l2 = [1] -> extra carry of
     * one at the end\ generate random nodes in linkedList
     */
    // Give a dummyHead ->ListNode dumHead = new ListNode(0)
    ListNode dumHead = new ListNode(0);
    // result = dumHead -> resultList
    // use p1 and p2 as pointers in L1 and L2
    ListNode p1 = l1, p2 = l2, cur = dumHead;
    // use cur as current pointer to point to next node in resultList
    // set carry = 0
    int carry = 0;
    // iterate over L1 and L2 until both of them has no node
    while (p1 != null || p2 != null) {
      // get values of p1 and p2
      int x1 = (p1 != null) ? p1.val : 0;
      int x2 = (p2 != null) ? p2.val : 0;
      // calculate sum = p1 + p2 + carry
      int sum = carry + x1 + x2;
      // carry = sum/10 as new carry value
      carry = sum / 10;
      // create new node(sum%10) for cur point to in resultList
      cur.next = new ListNode(sum % 10);
      cur = cur.next;
      // if p1 still has node -> p1 moves to next node in l1
      if (p1 != null)
        p1 = p1.next;
      // if p2 still has node -> p2 moves to next node in l2
      if (p2 != null)
        p2 = p2.next;
    }
    // if carry is larger than 0 -> add new node(carry) in result list
    if (carry > 0)
      cur.next = new ListNode(carry);
    // return dumHead.next as resultList
    return dumHead.next;
  }
}
